### PR TITLE
CEDARS hourly output post processing scripts update (New Feature)

### PR DESCRIPTION
**Pull request by lyt5410**
# Pull Request (PR) Description
**NOTE: SUBMIT A PLAN TO DEER SUPPORT PRIOR TO WORKING ON CODE FOR A MEASURE, FEATURE, OR BUG TO BE CONNECTED WITH OTHERS WORKING ON THE SAME ISSUE OR TO RECEIVE FEEDBACK ON YOUR PROPOSED METHOD**

New Feature Summary

This is the update to 4 existing post-processing python scripts and accompanying data tables in the folder _scripts\data transformation_ to produce hourly loadshape with CEDARS desired format in addition to their original functionality.

Updated the following scripts:
- SFm.py
- MFm.py
- DMo.py
- Com.py

Additional output table(s) that will be created by the update:
- CEDARS_long_ls_SFm.zip
- CEDARS_long_ls_MFm.zip
- CEDARS_long_ls_DMo.zip
- CEDARS_long_ls_Com.zip

Related data tables updates:
- normalizing unit related updates in DEER_EnergyPlus_Modelkit_Measure_list_working.xlsx
- normalizing unit related updates in Normunits.xlsx

These were developed using the following test run cases in the Modelkit system:

For SFm: 
**residential measures\SWBE013-01 Efficient Doors\SWBE013-01 Efficient Doors_SFm_1975**
- climate zones: CZ08, CZ09 
- cohort: SFm&1&rDXHP&Ex&BldgShell, SFm&2&rDXHP&Ex&BldgShell
- cases: RB-BldgShell-Door-Ex, RB-BldgShell-addDoorEff

**residential measures\SWBE013-01 Efficient Doors\SWBE013-01 Efficient Doors_SFm_1985**
- climate zones: CZ10, CZ11
- cohort: SFm&1&rDXHP&Ex&BldgShell, SFm&2&rDXHP&Ex&BldgShell
- cases: RB-BldgShell-Door-Ex, RB-BldgShell-addDoorEff

For MFm:
**residential measures\SWBE013-01 Efficient Doors\SWBE013-01 Efficient Doors_MFm_Ex**
- climate zones: CZ02, CZ03, CZ04
- cohort: MFm&0&rDXGF&Ex&BldgShell, MFm&0&rDXHP&Ex&BldgShell
- cases: RB-BldgShell-Door-Ex, RB-BldgShell-addDoorEff

For DMo:
**residential measures\SWBE013-01 Efficient Doors\SWBE013-01 Efficient Doors_DMo**
- climate zones: CZ02, CZ03
- cohort: DMo&0&rDXGF&Ex&BldgShell, DMo&0&rDXHP&Ex&BldgShell
- cases: RB-BldgShell-Door-Ex, RB-BldgShell-addDoorEff

For Com:
**commercial measures\SWHC024-06 Fan Belt\SWHC024-06 Fan Belt_Ex**
- climate zones: CZ02, CZ03
- cohort: (all available bldgtype cohorts)
- cases: NE-HVAC-FanBelt-SynchVBelt-Msr, NE-HVAC-FanBelt-VBelt-Std

### PR Author
- [x] Make sure the PR branch is up to date with main branch at the time of the PR submission
- [x] Craft a succinct title that effectively encapsulates the essence of the pull request, providing a general overview of the proposed changes.
- [x] Label the PR with at least one of the following: New Measure, Bug, or Feature.
- [x] Provide a concise description of the measure, bug, or feature. Submit one PR per measure.
- [ ] For a new measure, attach a workbook named DEER_EnergyPlus_Modelkit_Measure_list_working.xlsx, containing only rows used for post-processing the measure.
- [x] Add comments in the code when necessary to facilitate the review process.
- [x] Add a comment before the added code, including the author's full name, company, and specifying if it's a bug fix, new measure, or feature.
- [x] For a new feature or bug, demonstrate the impact on energy consumption for selected cases with justification using plots and descriptions.
- [ ] For a new measure, add a summary table showing total energy consumption per simulated case.
### PR Reviewer
- [ ] Conduct a thorough code review.
- [ ] If the branch is behind the main, merge the branch locally to check for potential conflicts.
- [ ] If a bug, locally reproduce it and compare energy consumptions before and after.
- [ ] Explore creative ways to stress-test the code.
- [ ] Locally check the error file and other outputs.

### CODE REVIEW

**Comment by lyt5410** [comment]
Update branch up-to-date with respect to current main, resolved conflicts

**Comment by nfette** [comment]
Here are some notes from testing with residential measure SWHC04-08.

- SWHC049-08 DMo: The script finished and created the output files.
- The script takes around 7 minutes to run for DMo building type.

**Comment by nfette** [comment]
Here are some notes from testing with commercial measure SWHC046-06.

- After around 15-30 minutes, I walked away. Later, the script failed to finish due to memory at line 530 inside FOR loop from line 501 (see error message below). I was running it on a Windows virtual machine with 16 GB memory. I see a lot of large dataframe variables in memory.
- I switched to a memory-optimized virtual machine with 32 GB memory. The script spends a lot of time running through the FOR loop around line 501. After 135 minutes, the script reached "column 2930 long format loaded" (approx 21 models/minute). After 480 minutes, it reached "col 7600 long format loaded" (approx 16 models/minute) and Python is using as much as 21 GB memory. After around 540 minutes, the script crashed with another memory error at line 629.
- This seems very slow. The script only utilizes 1 CPU core, and is inefficient in applying transformations to data one file at a time, and continuously concatenating to large dataframes. There are 8096 models for this measure. Assuming the script is processing models at a rate of 15-16 models/minute, the script would take over 8-9 hours to run. This is a concern because the script does not save intermediate progress, so any error would lose all progress, and we also tend to run simulations on a virtual machine with many CPU cores, so it is inconvenient or expensive to run a script for so long.

**Memory crash 1 at line 501**
```
#12/22/2025 CEDARS Hourly consumption output reformatting
# use the hourly data before long2wide pivot transform ...
reformatting hourly data for CEDARS loadshape format..
col 0 long format loaded.
col 1 long format loaded.
col 2 long format loaded.
...
col 5417 long format loaded.
col 5418 long format loaded.
col 5419 long format loaded.
---------------------------------------------------------------------------
MemoryError                               Traceback (most recent call last)
e:\CEDARS_long_output_update\scripts\data transformation\Com.py in <module>
     36     hrly_mapped['RunPeriod Start Day'] = col_names[4]
     37 
---> 38     converted_long_df = pd.concat([converted_long_df, hrly_mapped])
     39 
     40     print(f"col {i} long format loaded.")

...

MemoryError: Unable to allocate 362. MiB for an array with shape (1, 47487960) and data type object
```

**Memory crash 2 at line 629**
```
#rearrange / true-up columns ...
---------------------------------------------------------------------------
MemoryError                               Traceback (most recent call last)
e:\CEDARS_long_output_update\scripts\data transformation\Com.py in <module>
     20 #final table fields round-up
     21 #note: UEC and Numunits omitted from draft long table in the final table
---> 22 df_long_final = df_long[['Sector', 'BldgType','BldgVint','BldgHVAC','BldgLoc','Normunit',
     23          'Type', 'Source Year', 'TechGroup', 'TechType','TechID',
     24          'Hour of Year','UEC','UECproportion']] 

...

MemoryError: Unable to allocate 2.64 GiB for an array with shape (5, 70920960) and data type float64
```

File: scripts/data transformation/SFm.py
Comment (sepideh-QEA) [inline scripts/data transformation/SFm.py]: This branch removed the `MODE_NEW_VINTAGE` guard that exists on `main`, but it still unconditionally remaps `BldgVint` from climate zone here. If `path_new` is used, New-vintage runs will be rewritten to `1975` or `1985`, which would corrupt both the exported CEDARS loadshapes and the downstream `current_msr_mat` / SQL joins.

On `main`, this remap only happens in the non-new-vintage path:
`if MODE_NEW_VINTAGE: pass`
`else:`
`    sim_annual_f['BldgVint'] = sim_annual_f['BldgLoc'].map(cz_vint_dict)`
`    sim_hourly_final['BldgVint'] = sim_hourly_final['BldgLoc'].map(cz_vint_dict)`

I think this needs to preserve `BldgVint='New'` when `path_new` is used.

File: scripts/data transformation/Com.py [hunk 1]
```diff
@@ -468,17 +477,193 @@ def end_use_rearrange(df_in):
...
+converted_long_df['Total_Elec_Consumption'] = converted_long_df['Total_Elec_Consumption']/3600000
+
+
+# %%
+#modify this part for CEDARS
+#measure specific normalizing units table
+df_numunits = df_normunits[df_normunits['Normunit']==normunit]
+#finalize normalizing unit and num based on measure
+
+if len(df_numunits) == 1:
```
Comment (sepideh-QEA) [inline scripts/data transformation/Com.py]: When there is exactly one normunit value, the code sets numunits, but numunit_lookup is never defined. The later try block always references numunit_lookup, so single-value normunits hit the exception path and get exported as Normunit='Each' with Numunits=1. That silently produces incorrect UEC values instead of using the actual normalization from Normunits.xlsx.

File: scripts/data transformation/MFm.py [hunk 2]
```diff
@@ -340,30 +352,197 @@ def end_use_rearrange(df_in):
...
     numunits = list(numunits_vals['Value'])[0]
+elif len(df_numunits) > 1:
+    #for SEER rated ACHP, CZ specific Cap-Ton
+    normunit = df_numunits['Normunit'].unique()[0]
+    numunits = df_numunits.set_index('BldgLoc')['Value'].to_dict() #numunit is a dictionary
+    print(f'CZ-dependent numunits for this normalizing unit {normunit}')
 elif (measure_name == 'Wall Insulation') or (measure_name == 'Ceiling Insulation') or (measure_name == 'Windows'):
     numunits = list(numunits_vals[numunits_vals['Msr'] == measure_name]['Value'])[0]
+    print(f'normunit is {normunit}, numunits is {numunits}.')
 elif measure_name == 'PTAC / PTHP':
```
Comment (sepideh-QEA) [inline scripts/data transformation/MFm.py]: This branch changed numunits_vals to carry BldgLoc instead of CZ, but the PTAC/PTHP branch still indexes numunits_vals['CZ']. On main those names matched; here they do not, so existing PTAC/PTHP post-processing will fail before producing outputs

File: scripts/data transformation/DMo.py [hunk 3]
```diff
@@ -254,12 +264,16 @@ def end_use_rearrange(df_in):
     full_path = hrly_path + "/" + split_meta_cols_eu.iloc[i][0] + "/" + split_meta_cols_eu.iloc[i][1] + "/" + split_meta_cols_eu.iloc[i][2] + "/instance-var.csv"
     df = pd.read_csv(full_path, low_memory=False)
     
+    #3/3/2026 update, extract RunPeriod Start Day from IDF file for a particular simulation
+    idf_path = hrly_path + "/" + split_meta_cols_eu.iloc[i][0] + "/" + split_meta_cols_eu.iloc[i][1] + "/" + split_meta_cols_eu.iloc[i][2] + "/instance.idf"
+    runperiod_start_day = helper_functions.get_runperiod_start_day(idf_path)
+
     #extract the last column (the total elec hrly profile)
```
Comment (sepideh-QEA) [inline scripts/data transformation/DMo.py]: On main, these scripts explicitly read Electricity:Facility [J](Hourly). This PR switches to df.iloc[:, -1], which makes the export depend on whatever column happens to be last in instance-var.csv. If the report variable order changes or another hourly series is appended, the CEDARS loadshape and the existing hourly post-processing will silently use the wrong signal.

**Comment by sepideh-QEA** [comment]
@nfette Thanks for the testing notes. I think the runtime and memory behavior you saw point to a real scalability issue, not just a VM-specific problem. The current implementation appears to build very large long-format dataframes by repeated concatenation and then materializes another large final dataframe, so the behavior you observed looks plausible for large commercial measures.

File: scripts/data transformation/Com.py (hunk 1 above)
Comment (sepideh-QEA) [reply scripts/data transformation/Com.py]: also from @Justin-DeBlois: measures with multiple intended normunits (example cited: Len-ft_mt vs Len-ft_lt in the mapping workbook) hit this path, and because `numunit_lookup` isn’t defined the `try/except` silently defaults to `Normunit='Each'` / `Numunits=1`. That produces “valid-looking” output but with incorrect normalization metadata/UEC. Can we fix this by (1) defining `numunit_lookup` for the single-value case and (2) replacing the broad `except` with explicit validation + a hard error when normunit/numunits can’t be determined?

File: scripts/data transformation/Com.py [hunk 4]
```diff
@@ -468,17 +477,193 @@ def end_use_rearrange(df_in):
...
+    "Tuesday": 2013, #2013 or 2019 could be used
+    "Wednesday": 2020, #Basis for 2022/2021 electric ACCs
+    "Thursday": 2009, #Per CEC's Nonres/MFm ACM Reference Manual
+    "Friday": 2010, #2016 is Friday but a leap year, so this should be either 2010 or 2021
+    "Saturday": 2011, #Next Saturday option is 2022 because it is skipped between 2016 and 2017 because 2016 is a leap year
+    "Sunday": 2017 #2012 is a leap year, suggest using 2017
+}
+
+df_long['Sector'] = 'Com' #this is Com script, so Sector = Com
+df_long['Type'] = 'Whole Building'
```
Comment (sepideh-QEA) [inline scripts/data transformation/Com.py]: Type is currently hard-coded to Whole Building because the script is exporting whole-building electricity (facility) only. End-use-specific Type would require end-use hourly outputs (lights/HVAC/etc.) plus additional mapping/aggregation logic, which isn’t available in the current post-processing inputs and would expand file size/memory load. Suggest keeping Whole Building for PR #133; we can consider an end-use extension later if/when end-use series are included in the simulation outputs.

**Comment by lyt5410** [comment]
> Here are some notes from testing with commercial measure SWHC046-06.
> 
> * After around 15-30 minutes, I walked away. Later, the script failed to finish due to memory at line 530 inside FOR loop from line 501 (see error message below). I was running it on a Windows virtual machine with 16 GB memory. I see a lot of large dataframe variables in memory.
> * I switched to a memory-optimized virtual machine with 32 GB memory. The script spends a lot of time running through the FOR loop around line 501. After 135 minutes, the script reached "column 2930 long format loaded" (approx 21 models/minute). After 480 minutes, it reached "col 7600 long format loaded" (approx 16 models/minute) and Python is using as much as 21 GB memory. After around 540 minutes, the script crashed with another memory error at line 629.
> * This seems very slow. The script only utilizes 1 CPU core, and is inefficient in applying transformations to data one file at a time, and continuously concatenating to large dataframes. There are 8096 models for this measure. Assuming the script is processing models at a rate of 15-16 models/minute, the script would take over 8-9 hours to run. This is a concern because the script does not save intermediate progress, so any error would lose all progress, and we also tend to run simulations on a virtual machine with many CPU cores, so it is inconvenient or expensive to run a script for so long.
> 
> **Memory crash 1 at line 501**
> 
> ```
> #12/22/2025 CEDARS Hourly consumption output reformatting
> # use the hourly data before long2wide pivot transform ...
> reformatting hourly data for CEDARS loadshape format..
> col 0 long format loaded.
> col 1 long format loaded.
> col 2 long format loaded.
> ...
> col 5417 long format loaded.
> col 5418 long format loaded.
> col 5419 long format loaded.
> ---------------------------------------------------------------------------
> MemoryError                               Traceback (most recent call last)
> e:\CEDARS_long_output_update\scripts\data transformation\Com.py in <module>
>      36     hrly_mapped['RunPeriod Start Day'] = col_names[4]
>      37 
> ---> 38     converted_long_df = pd.concat([converted_long_df, hrly_mapped])
>      39 
>      40     print(f"col {i} long format loaded.")
> 
> ...
> 
> MemoryError: Unable to allocate 362. MiB for an array with shape (1, 47487960) and data type object
> ```
> 
> **Memory crash 2 at line 629**
> 
> ```
> #rearrange / true-up columns ...
> ---------------------------------------------------------------------------
> MemoryError                               Traceback (most recent call last)
> e:\CEDARS_long_output_update\scripts\data transformation\Com.py in <module>
>      20 #final table fields round-up
>      21 #note: UEC and Numunits omitted from draft long table in the final table
> ---> 22 df_long_final = df_long[['Sector', 'BldgType','BldgVint','BldgHVAC','BldgLoc','Normunit',
>      23          'Type', 'Source Year', 'TechGroup', 'TechType','TechID',
>      24          'Hour of Year','UEC','UECproportion']] 
> 
> ...
> 
> MemoryError: Unable to allocate 2.64 GiB for an array with shape (5, 70920960) and data type float64
> ```

@nfette I made some revamps to the hourly data parsing methods (both the 365x24 step and the CEDARS long 8760 step) in just the commercial script **Com.py** such that it will no longer loop over multiple pd.DataFrame calls and only loops data over numpy arrays + dictionaries then do 1 dataframe creation after all data has been parsed, I think this should reduce the memory cost. It was a rookie mistake that I have overlooked for scalability - thanks for spending the time to test this out.

Do you mind testing your commercial cohort run again with my updates and let me know how it goes? I assume you still have the simulation results set on your machine.

If this solves the issue, I will apply this to the other 3 scripts. If not, we can work something out.

File: scripts/data transformation/Com.py [hunk 5]
```diff
@@ -468,17 +506,224 @@ def end_use_rearrange(df_in):
...
+    long_data["Measure Group Name"].append(np.repeat(parts[1], N_HOURS))
+    long_data["TechID"].append(np.repeat(parts[2], N_HOURS))
+    long_data["file"].append(np.repeat(parts[3], N_HOURS))
+    long_data["RunPeriod Start Day"].append(np.repeat(parts[4], N_HOURS))
+
+    print(f"col {i} long format loaded.")
+
+#build dataframe once
+final_data = {k: np.concatenate(v) for k, v in long_data.items()}
+converted_long_df = pd.DataFrame(final_data)
```
Comment (nfette) [inline scripts/data transformation/Com.py:666]: @lyt5410 Everything up to this point runs much more quickly now, thanks. At line 591, I am running out of memory (32 GB RAM). But there might not be a great way to avoid this. With 128 GB RAM, the script finished without error and took about 25 minutes.

Comment (lyt5410) [reply scripts/data transformation/Com.py:666]: @sepideh-QEA See my updates and @nfette 's comments on the fix. The main memory issue is solved but with the caveat that a 128 GB machine is required to run a large full Commercial run. Is this an acceptable fix? Or do we want to further develop a work-around for this (such as slicing up the data to output CSVs per building type, using Parquet to store data, etc.) so that any PA/measure developers can do this process with a full commercial run as big as @nfette 's test case here with a 32GB machine.

@nfette can I get a sense of how many permutations/simulations is your test batch run?

In the meantime, I will be fixing the other issues (normunits, etc.).

**Comment by lyt5410** [comment]
In the process of QC'ing scripts and making sure everything connects, I noticed that the Fan Belt measure folder (SWHC024-06 Fan Belt, the one I was testing for Com script) has some inconsistent labeling on the BldgHVAC component - "rDXHP" instead of "cDXHP" despite being a commercial measure. The scripts will still run but this discrepant descriptor/labelling is carried along the way, which is why I did not note this earlier. Maybe consider adding a check somewhere along the data process pipeline in the future.
image

File: scripts/data transformation/Com.py (hunk 5 above)
Comment (nfette) [reply scripts/data transformation/Com.py:666]: I was testing with commercial measure SWHC046-06, which has 8096 model instances.

I think we can work with the proposed fix. During production, we might try to run the script with a small batch of models to make sure there are no errors, then proceed to run with all models.

File: scripts/data transformation/SFm.py
Comment (lyt5410) [reply scripts/data transformation/SFm.py]: updated SFm.py to made sure it keeps this BldgVint guard.

File: scripts/data transformation/Com.py (hunk 1 above)
Comment (lyt5410) [reply scripts/data transformation/Com.py]: for Com.py, I removed sections on CEDARS updated that involved normalizing units.

File: scripts/data transformation/MFm.py (hunk 2 above)
Comment (lyt5410) [reply scripts/data transformation/MFm.py]: fixed this lookup to refer to 'BldgLoc' instead of 'CZ', and moved this section out of CEDARS update portion (to after it is done), since normalizing unit related code are removed from the CEDARS update portion.

File: scripts/data transformation/Com.py (hunk 5 above)
Comment (lyt5410) [reply scripts/data transformation/Com.py:666]: Sounds good. I made this update to the Res scripts as well for consistency.

File: scripts/data transformation/Com.py
[the remark this answers is not in the document]
Comment (sepideh-QEA) [reply scripts/data transformation/Com.py]: @lyt5410 this hasn't been resolved. There is no explicit validation right after reading Normunits.xlsx to confirm total_area_m2 exists on the Com sheet. So if a user has an older/different Normunits.xlsx (or the wrong sheet layout), they can still get the same late-stage failure: CEDARS zip created first, then crash here before writing the 3 standard DEER CSVs.

Can you add a clear check like “if total_area_m2 not in df_normunits.columns: raise ValueError(…)” and note that the PR’s updated Normunits.xlsx is required for the commercial path.

**Comment by sepideh-QEA** [comment]
@lyt5410 TRC retested PR #133 (5/12/2026) on behalf of PG&E. They report Com.py still fails before producing the standard DEER outputs due to the area_lookup step expecting total_area_m2 in df_normunits:

area_lookup = df_normunits[df_normunits['Normunit']=='Area-ft2-BA'][['BldgType','total_area_m2']]

In their run (SWCR001 refrigeration, Gro prototype), df_normunits did not contain total_area_m2 (they only see BldgType, BldgLoc, Normunit, Value, BldgVint), so the script stops and the DEER CSVs aren’t generated (even if the CEDARS zip is created earlier).

Please either (1) switch this logic to use the existing Value column for Area-ft2-BA (and convert units if needed), or (2) add explicit validation right after reading Normunits.xlsx with a clear error message telling users which Normunits file/schema is required. see screenshots here.
[Energy Plus Post Processing Script for Load Shapes_TRC_5-13-26.docx](https://github.com/user-attachments/files/27729048/Energy.Plus.Post.Processing.Script.for.Load.Shapes_TRC_5-13-26.docx)

**Comment by lyt5410** [comment]
> @lyt5410 TRC retested PR #133 (5/12/2026) on behalf of PG&E. They report Com.py still fails before producing the standard DEER outputs due to the area_lookup step expecting total_area_m2 in df_normunits:
> 
> area_lookup = df_normunits[df_normunits['Normunit']=='Area-ft2-BA'][['BldgType','total_area_m2']]
> 
> In their run (SWCR001 refrigeration, Gro prototype), df_normunits did not contain total_area_m2 (they only see BldgType, BldgLoc, Normunit, Value, BldgVint), so the script stops and the DEER CSVs aren’t generated (even if the CEDARS zip is created earlier).
> 
> Please either (1) switch this logic to use the existing Value column for Area-ft2-BA (and convert units if needed), or (2) add explicit validation right after reading Normunits.xlsx with a clear error message telling users which Normunits file/schema is required. see screenshots here. [Energy Plus Post Processing Script for Load Shapes_TRC_5-13-26.docx](https://github.com/user-attachments/files/27729048/Energy.Plus.Post.Processing.Script.for.Load.Shapes_TRC_5-13-26.docx)

Updated area_lookup to be a dictionary of bldgtype-specific area values. The only objective of that step was to add the modelled area to the table.

**Comment by sepideh-QEA** [comment]
@lyt5410 TRC (@Justin-DeBlois) retested after the revision and confirms the script now runs, but they found a logic bug in the new normunit_missing handling: for normunit = 'Len-ft_mt' / 'Len-ft_lt', the code is incorrectly evaluating normunit_missing = True because it uses list(normunit) (which splits the string into characters). Can you update the check to compare strings against df_normunits['Normunit'] values directly (e.g., if normunit not in set(df_normunits['Normunit'].astype(str))) and avoid character-list logic?

**Comment by lyt5410** [comment]
> @lyt5410 TRC (@Justin-DeBlois) retested after the revision and confirms the script now runs, but they found a logic bug in the new normunit_missing handling: for normunit = 'Len-ft_mt' / 'Len-ft_lt', the code is incorrectly evaluating normunit_missing = True because it uses list(normunit) (which splits the string into characters). Can you update the check to compare strings against df_normunits['Normunit'] values directly (e.g., if normunit not in set(df_normunits['Normunit'].astype(str))) and avoid character-list logic?

during our call today (5/15/16 11am), looks like this issue is resolved with the latest commit.

**Comment by Justin-DeBlois** [comment]
We did some spot checks using SWCR001 and it looks like our comments have been resolved. All good on my end.

**Comment by lyt5410** [comment]
I did the regression testing (main branch vs. PR133 branch) after this whole set of updates. Here are the results:

Measure / Permutation tested:
**commercial measures\SWHC024-06 Fan Belt\SWHC024-06 Fan Belt_Ex
climate zones: CZ02, CZ03
cohort: (all available bldgtype cohorts)
cases: NE-HVAC-FanBelt-SynchVBelt-Msr, NE-HVAC-FanBelt-VBelt-Std**

test results files here:
[Com_regression_test_051526.zip](https://github.com/user-attachments/files/27822264/Com_regression_test_051526.zip)


#### Branch PR#133:
Tested the measure SWHC024-06 Fan Belt
Com.py script works to produce the DEER impact required outputs
**_the "measarea" field is in square-feet (ft2)_** 

#### Main branch:
Tested the measure SWHC024-06 Fan Belt
Com.py script works to produce the DEER impact required outputs if Normunits.xlsx is using the updated version. (existing main branch Normunits.xlsx table did not have Cap-Tons unit listed, hence it produced blank sim_annual tables with the main-branch Normunits.xlsx because the normunit lookup would return nothing)
**_the "measarea" field is in square-meters (m2)_**

The only difference between main branch run and PR#133 run is the measarea
image

No changes are made in the SQL pipeline, but do note where the field "measarea" is used in the SQL pipeline. It looked like "measarea" is one of the numeric values that gets weighted downstream and carried over to impact tables. However, I have a limited knowledge on how that field gets used beyond that point when it is in DEER impacts format, but since the measarea field was part of the original eQuest/MASControl3 workflow output, we kept that field for consistency. (My guess is that it may be used as a way to calculate EUI downstream.)

Normunit and numunits are behaving as expected, since right now it appears as the placeholder value 10 for Cap-Tons. It is up to the measure developer to update this value to reflect the actual measure. And the savings would reflect that.

As we have discussed on the call, for standardization purposes, the correct numunit values should be updated within the Normunits.xlsx table (and the corresponding normalizing unit type should be specified in the starting workbook "DEER_EnergyPlus_Modelkit_Measure_list_working_*.xlsx". The Com.py script should now have error validations to break the run if these files are missing/mismatch.

Measure developers can also choose the replace the numunit values as an external step if the current script functionality doesn't cover edge cases, for example if a normalizing unit varies based on climate-zone (e.g. I had a brief exchange with Shaojie (@wsjsxn) from Energy Solutions on the new script testing, that was a workaround he mentioned)

Shaojie and I made sure his test case worked without issue as well.

**Comment by lyt5410** [comment]
Additional context that was discussed during today's call. Putting this here just to document.

**Normunit edge cases and current limitations**
The current Com.py normalizing unit handling process (including the original implementation on the main branch) is not fully robust. In particular, it currently does not support workflows where more than one unique `NormUnit` is present within a single starting workbook (`DEER_EnergyPlus_Modelkit_Measure_list_working_*.xlsx`).

There are also additional edge cases (beyond my current knowledge and scope) that could arise in future measure development. These may require further enhancements to the normalization logic and feedbacks from measure developers, depending on how measures evolve. Examples include:

- Normalizing unit varies by building type *and* climate zone  
  *(e.g., climate-zone-specific cooling capacity values; A 'BldgLoc' field and makeshift example NormUnits are provided in the Normunits.xlsx table, but needs further development for production)*  

- Multiple normalizing units within a single measure  
  *(e.g., `len-ft_mt` vs. `len-ft_lt` for anti-sweat heater controls)*  

- Normalizing unit varies based on additional prototype characteristics beyond building type or climate zone  

- Normalizing unit is a constant scalar value  
  *(e.g., `Each`, `reducedkW`; these cases are expected to continue working with minimal changes)*  

Overall, the current implementation assumes a simplified one-NormUnit-per-batch structure, and will require refactoring to support more complex or heterogeneous normalization schemes in the future.

**Comment by lyt5410** [comment]
Added some update to the Readme regarding the normalizing unit topic.
